File: elixir/raindrops/raindrops.exs
defmodule Raindrops do
  @factors [3, 5, 7]

  @doc """
  Returns a string based on raindrop factors.

  - If the number contains 3 as a prime factor, output 'Pling'.
  - If the number contains 5 as a prime factor, output 'Plang'.
  - If the number contains 7 as a prime factor, output 'Plong'.
  - If the number does not contain 3, 5, or 7 as a prime factor,
    just pass the number's digits straight through.
  """
  @spec convert(pos_integer) :: String.t()
  def convert(number) do
    raindrops = for f <- @factors, rem(number, f) == 0 do
      sound(f)
    end

    if raindrops == [] do
        number
    else
        raindrops
    end
    |> to_string()
  end

  defp sound(3), do: "Pling"
  defp sound(5), do: "Plang"
  defp sound(7), do: "Plong"
end

File: erlang/raindrops/src/raindrops.erl
-module(raindrops).

-define(FACTORS, [3, 5, 7]).

-export([convert/1, test_version/0]).

%% API

-spec convert(integer()) -> string().
convert(Number) ->
    case [sound(F) || F <- ?FACTORS, Number rem F == 0] of
        [] ->
            integer_to_list(Number);
        Sounds ->
            lists:flatten(Sounds)
    end.

-spec test_version() -> integer().
test_version() ->
    1.

%% Internal

sound(3) ->
    "Pling";
sound(5) ->
    "Plang";
sound(7) ->
    "Plong".
